cvmfs/cache_plugin/cvmfs_cache_posix.cc

```diff
@@ -0,0 +1,415 @@
+    int fd = cache_mgr->Open(blessed_object);
+    if (fd < 0) {
+      return CVMCACHE_STATUS_NOENTRY;
+    }
+    object.fd = fd;
+    object.size = cache_mgr->GetSize(fd);
+    object.refcnt = 0;
+    pinned_size += object.size;
+  } else if (object.refcnt + change_by < 0) {
+    return CVMCACHE_STATUS_BADCOUNT;
+  }
+
+  object.refcnt += change_by;
+  if (object.refcnt == 0) {
+    if (cache_mgr->Close(object.fd)) {
+      return CVMCACHE_STATUS_IOERR;
+    }
+    pinned_size -= object.size;
+    opened_objects->Erase(*id);
+  } else {
+    opened_objects->Insert(*id, object);
+  }
+  return CVMCACHE_STATUS_OK;
+}
+
+// Only gives info for opened objects.
+// Should be fine, since cvmfs only requests info for opened objects.
+static int posix_obj_info(struct cvmcache_hash *id,
+                          struct cvmcache_object_info *info) {
+  CacheObject object;
+  if (!opened_objects->Lookup(*id, &object)) {
+    return CVMCACHE_STATUS_NOENTRY;
+  }
+  info->id = *id;
+  info->size = object.size;
+  return CVMCACHE_STATUS_OK;
+}
+
+static int posix_pread(struct cvmcache_hash *id, uint64_t offset,
+                       uint32_t *size, unsigned char *buffer) {
+  CacheObject object;
+  if (!opened_objects->Lookup(*id, &object)) {
+    return CVMCACHE_STATUS_NOENTRY;
+  }
+  if (offset > object.size) {
+    return CVMCACHE_STATUS_OUTOFBOUNDS;
+  }
+  int64_t bytes_read = cache_mgr->Pread(object.fd, buffer, *size, offset);
+  if (bytes_read < 0) {
+    return CVMCACHE_STATUS_IOERR;
+  }
+  *size = (uint32_t) bytes_read;
+  return CVMCACHE_STATUS_OK;
+}
+
+static int posix_start_txn(struct cvmcache_hash *id,
+                           uint64_t txn_id,
+                           struct cvmcache_object_info *info) {
+  void *txn = malloc(cache_mgr->SizeOfTxn());  // cachemgr deletes in commit_txn
+  int fd = cache_mgr->StartTxn(Chash2Cpphash(id), info->size, txn);
+  if (fd < 0) {
+    return CVMCACHE_STATUS_IOERR;
+  }
+  Txn transaction;
+  transaction.fd = fd;
+  transaction.hash = *id;
+  transaction.txn = txn;
+  transaction.size = info->size;
+  transaction.type = info->type;
+  transactions->Insert(txn_id, transaction);
+
+  CacheManager::ObjectInfo object_info;
+  if (info->type == CVMCACHE_OBJECT_CATALOG) {
+    object_info.type = CacheManager::kTypeCatalog;
+  } else if (info->type == CVMCACHE_OBJECT_VOLATILE) {
+    object_info.type = CacheManager::kTypeVolatile;
+  } else {
+    object_info.type = CacheManager::kTypeRegular;
+  }
+  if (info->description) {
+    object_info.description = info->description;
+  }
+  cache_mgr->CtrlTxn(object_info, 0, txn);
+  return CVMCACHE_STATUS_OK;
+}
+
+static int posix_write_txn(uint64_t txn_id, unsigned char *buffer,
+                           uint32_t size) {
+  Txn transaction;
+  if (!transactions->Lookup(txn_id, &transaction)) {
+    return CVMCACHE_STATUS_NOENTRY;
+  }
+  int64_t bytes_written = cache_mgr->Write(buffer, size, transaction.txn);
+  if (bytes_written < 0) {
```

Comment:
Shouldn't it be `(bytes_written >= 0) && (static_cast<uint32_t>(bytes_written) == size)`?

Reply:
I checked functions PosixCacheManager::Write and PosixCacheManager::Flush and the Write function should either return negative value or `size`. Am I missing something? Should I do it as the comment says anyway?

Reply:
Ok.  Let's still be defensive and explicitly check for size.